boa_engine/src/error.rs

```diff
@@ -0,0 +1,787 @@
+    /// or `None` otherwise.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use boa_engine::{JsError, JsNativeError};
+    /// let error: JsError = JsNativeError::reference()
+    ///     .with_message("variable not found!")
+    ///     .into();
+    ///
+    /// assert!(error.as_opaque().is_none());
+    ///
+    /// let error = JsError::from_opaque(256u32.into());
+    ///
+    /// assert!(error.as_opaque().is_some());
+    /// ```
+    pub fn as_opaque(&self) -> Option<&JsValue> {
+        match self.inner {
+            Repr::Native(_) => None,
+            Repr::Opaque(ref v) => Some(v),
+        }
+    }
+
+    /// Gets the inner [`JsNativeError`] if the error is a native
+    /// error, or `None` otherwise.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use boa_engine::{JsError, JsNativeError, JsValue};
+    /// let error: JsError = JsNativeError::error().with_message("Unknown error").into();
+    ///
+    /// assert!(error.as_native().is_some());
+    ///
+    /// let error = JsError::from_opaque(JsValue::undefined().into());
+    ///
+    /// assert!(error.as_native().is_none());
+    /// ```
+    pub fn as_native(&self) -> Option<&JsNativeError> {
+        match self.inner {
+            Repr::Native(ref e) => Some(e),
+            Repr::Opaque(_) => None,
+        }
+    }
+}
+
+impl From<parser::ParseError> for JsError {
+    fn from(err: parser::ParseError) -> Self {
+        Self::from(JsNativeError::from(err))
+    }
+}
+
+impl From<JsNativeError> for JsError {
+    fn from(error: JsNativeError) -> Self {
+        Self {
+            inner: Repr::Native(error),
+        }
+    }
+}
+
+impl Display for JsError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match &self.inner {
+            Repr::Native(e) => e.fmt(f),
+            Repr::Opaque(v) => v.display().fmt(f),
+        }
+    }
+}
+
+/// Native representation of an ideal `Error` object from Javascript.
+///
+/// This representation is more space efficient than its [`JsObject`] equivalent,
+/// since it doesn't need to create a whole new `JsObject` to be instantiated.
+/// Prefer using this over [`JsError`] when you don't need to throw
+/// plain [`JsValue`]s as errors, or when you need to inspect the error type
+/// of a `JsError`.
+///
+/// # Examples
+///
+/// ```rust
+/// # use boa_engine::{JsNativeError, JsNativeErrorKind};
+/// let native_error = JsNativeError::uri().with_message("cannot decode uri");
+///
+/// match native_error.kind {
+///     JsNativeErrorKind::Uri => { /* handle URI error*/ }
+///     _ => unreachable!(),
+/// }
+///
+/// assert_eq!(native_error.message(), "cannot decode uri");
+/// ```
+#[derive(Debug, Clone, Trace, Finalize, Error)]
+#[error("{kind}: {message}")]
+pub struct JsNativeError {
+    /// The kind of native error (e.g. `TypeError`, `SyntaxError`, etc.)
+    pub kind: JsNativeErrorKind,
+    #[unsafe_ignore_trace]
+    message: LocalStr,
```

Comment:
Do we really need this now? I understand this is more space efficient and has performance benefits, but we are adding 2 new dependencies, so it might make sense to think about this thoroughly and implement them system-wide. We also talked about small_str, and we decided in the end to intern strings, so that might be another option.

Reply:
Hmm yeah, maybe I prematurely optimized this. I'll replace it with a simple `Rc<str>`.